sky/skylet/providers/gcp/config.py

```diff
@@ -504,24 +504,45 @@ def _check_firewall_rules(vpc_name, config, compute):
         return False
     effective_rules = response["firewalls"]
 
-    def _get_refined_rule(rule):
-        KEY_TO_COMPARE = {"sourceRanges", "allowed", "direction"}
-        refined_rule = {}
-        for k in KEY_TO_COMPARE:
-            if k not in rule:
-                continue
-            if k == "allowed":
-                refined_rule[k] = sorted(rule[k], key=lambda x: x["IPProtocol"])
-            else:
-                refined_rule[k] = rule[k]
-        return refined_rule
-
-    required_rules = list(map(_get_refined_rule, required_rules))
-    effective_rules = list(map(_get_refined_rule, effective_rules))
-
-    for rule in required_rules:
-        if rule not in effective_rules:
+    def _merge_and_refine_rule(rules):
+        # Output: dict of (direction, source) -> {protocol -> ports}
+        tmp = {}
+        for rule in rules:
+            direction = rule.get("direction", "")
+            sources = rule.get("sourceRanges", [])
+            allowed = rule.get("allowed", [])
+            for source in sources:
+                tmp[(direction, source)] = tmp.get((direction, source), []) + allowed
+        source2rules = {}
```

Comment:
Can we have a type annotation for this? `Dict[Tuple[str, str], Dict[str, Set[int]]]`

Also, it would be nice to have a comment showing an example object of this dict.

Reply:
Yes, annotation and example added